File: backend/routes/analytics.py
from fastapi import APIRouter
from database import db

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


@router.get("/trends")
def trends():
    all_waste = db.get_all("waste_summary")
    grouped: dict = {}
    for r in all_waste:
        d = r.get("date", "")
        grouped[d] = grouped.get(d, 0) + (r.get("total_wasted_kg") or r.get("wasted_kg") or 0)
    result = [{"date": d, "waste": round(w, 2)} for d, w in grouped.items()]
    result.sort(key=lambda x: x["date"], reverse=True)
    return list(reversed(result[:7]))


@router.get("/waste-by-item")
def waste_by_item():
    all_waste = db.get_all("waste_summary")
    foods = db.get_all("food_items")
    by_item: dict = {}
    for w in all_waste:
        iid = w.get("item_id") or w.get("food_item_id")
        if iid not in by_item:
            food = next((f for f in foods if f["id"] == iid), None)
            by_item[iid] = {"name": food["name"] if food else "Unknown", "served": 0, "wasted": 0, "count": 0}
        by_item[iid]["served"] += w.get("total_served_kg") or w.get("served_kg") or 0
        by_item[iid]["wasted"] += w.get("total_wasted_kg") or w.get("wasted_kg") or 0
        by_item[iid]["count"] += 1
    result = []
    for iid, d in by_item.items():
        pct = round((d["wasted"] / d["served"]) * 100, 1) if d["served"] > 0 else 0
        result.append({"item_id": iid, **d, "waste_percent": pct, "waste_cost": round(d["wasted"] * 120)})
    result.sort(key=lambda x: x["waste_percent"], reverse=True)
    return {"success": True, "data": result}


@router.get("/waste-by-meal")
def waste_by_meal():
    all_waste = db.get_all("waste_summary")
    by_meal: dict = {}
    for w in all_waste:
        m = w.get("meal_type", "")
        if m not in by_meal:
            by_meal[m] = {"served": 0, "wasted": 0, "count": 0}
        by_meal[m]["served"] += w.get("total_served_kg") or w.get("served_kg") or 0
        by_meal[m]["wasted"] += w.get("total_wasted_kg") or w.get("wasted_kg") or 0
        by_meal[m]["count"] += 1
    result = []
    for meal, d in by_meal.items():
        pct = round((d["wasted"] / d["served"]) * 100, 1) if d["served"] > 0 else 0
        result.append({"meal_type": meal, **d, "waste_percent": pct})
    return {"success": True, "data": result}


@router.get("/sentiment")
def sentiment():
    reviews = db.get_all("reviews")
    total = len(reviews)
    positive = sum(1 for r in reviews if r.get("overall_rating", 0) >= 4)
    neutral = sum(1 for r in reviews if r.get("overall_rating", 0) == 3)
    negative = sum(1 for r in reviews if r.get("overall_rating", 0) <= 2)
    avg = round(sum(r.get("overall_rating", 0) for r in reviews) / total, 1) if total > 0 else 0
    return {"success": True, "data": {"total": total, "positive": positive, "neutral": neutral, "negative": negative, "avgRating": avg}}

File: routes/analytics.ts
// Analytics Routes
import { Express } from "express";

export function registerAnalyticsRoutes(app: Express, db: any) {
  // Waste trends (last 7 days)
  app.get("/api/analytics/trends", (req, res) => {
    try {
      const trends = db.prepare("SELECT date, SUM(wasted_kg) as waste FROM waste_summary GROUP BY date ORDER BY date DESC LIMIT 7").all();
      res.json(trends.reverse());
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  // Item-wise waste
  app.get("/api/analytics/waste-by-item", (req, res) => {
    try {
      const allWaste = db.prepare("SELECT * FROM waste_summary").all();
      const foods = db.prepare("SELECT * FROM food_items").all();
      const byItem: Record<number, { name: string; served: number; wasted: number; count: number }> = {};
      allWaste.forEach((w: any) => {
        const id = w.item_id || w.food_item_id;
        if (!byItem[id]) {
          const food = foods.find((f: any) => f.id === id);
          byItem[id] = { name: food?.name || 'Unknown', served: 0, wasted: 0, count: 0 };
        }
        byItem[id].served += w.total_served_kg || w.served_kg || 0;
        byItem[id].wasted += w.total_wasted_kg || w.wasted_kg || 0;
        byItem[id].count++;
      });
      const result = Object.entries(byItem).map(([id, data]) => ({
        item_id: parseInt(id), ...data,
        waste_percent: data.served > 0 ? Math.round((data.wasted / data.served) * 1000) / 10 : 0,
        waste_cost: Math.round(data.wasted * 120)
      })).sort((a, b) => b.waste_percent - a.waste_percent);
      res.json({ success: true, data: result });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  // Meal-wise comparison
  app.get("/api/analytics/waste-by-meal", (req, res) => {
    try {
      const allWaste = db.prepare("SELECT * FROM waste_summary").all();
      const byMeal: Record<string, { served: number; wasted: number; count: number }> = {};
      allWaste.forEach((w: any) => {
        const m = w.meal_type;
        if (!byMeal[m]) byMeal[m] = { served: 0, wasted: 0, count: 0 };
        byMeal[m].served += w.total_served_kg || w.served_kg || 0;
        byMeal[m].wasted += w.total_wasted_kg || w.wasted_kg || 0;
        byMeal[m].count++;
      });
      const result = Object.entries(byMeal).map(([meal, data]) => ({
        meal_type: meal, ...data, waste_percent: data.served > 0 ? Math.round((data.wasted / data.served) * 1000) / 10 : 0
      }));
      res.json({ success: true, data: result });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });

  // Review sentiment summary
  app.get("/api/analytics/sentiment", (req, res) => {
    try {
      const reviews = db.prepare("SELECT * FROM reviews ORDER BY id DESC").all();
      const positive = reviews.filter((r: any) => r.overall_rating >= 4).length;
      const neutral = reviews.filter((r: any) => r.overall_rating === 3).length;
      const negative = reviews.filter((r: any) => r.overall_rating <= 2).length;
      res.json({ success: true, data: { total: reviews.length, positive, neutral, negative, avgRating: reviews.length > 0 ? (reviews.reduce((s: number, r: any) => s + r.overall_rating, 0) / reviews.length).toFixed(1) : 0 } });
    } catch (err: any) {
      res.status(500).json({ success: false, error: err.message });
    }
  });
}
